handler/OpenIDLoginHandler.php

```diff
@@ -0,0 +1,302 @@
+
+        return false;
+    }
+
+    /**
+     * Used for legacy login in case of errors or other bad things.
+     *
+     * @param $args
+     * @param $request
+     */
+    function legacyLogin($args, $request)
+    {
+        $templateMgr = TemplateManager::getManager($request);
+        $this->_enableLegacyLogin($templateMgr, $request);
+        $templateMgr->assign('disableUserReg', true);
+
+        return $templateMgr->display('frontend/pages/userLogin.tpl');
+    }
+
+    /**
+     * Overwrites the default registration, because it is not needed anymore.
+     * User registration is done via OpenID provider.
+     *
+     * @param $args
+     * @param $request
+     */
+    function register($args, $request)
+    {
+        $this->index($args, $request);
+    }
+
+    /**
+     * Overwrites default signOut.
+     * Performs OJS logout and if logoutUrl is provided (e.g. Apple doesn't provide this url) it redirects to the oauth logout to delete session and tokens.
+     *
+     * @param $args
+     * @param $request
+     */
+    function signOut($args, $request)
+    {
+        if (Validation::isLoggedIn()) {
+            $plugin = PluginRegistry::getPlugin('generic', KEYCLOAK_PLUGIN_NAME);
+            $router = $request->getRouter();
+
+            $requestUser = $request->getUser();
+
+            $user = null;
+            $lastProvider = null;
+            if ($requestUser) {
+                $user = Repo::user()->get($requestUser->getId());
+
+                $lastProvider = $user->getData('openid::lastProvider');
+            }
+            
+            $context = Application::get()->getRequest()->getContext();
+            $contextId = ($context == null) ? 0 : $context->getId();
+            $settingsJson = $plugin->getSetting($contextId, 'openIDSettings');
+
+            if (isset($user)) {
+                $user->setData('openid::lastProvider', null);
+                Repo::user()->edit($user);
+            }
+
+            $sessionManager = SessionManager::getManager();
+            $session = $sessionManager->getUserSession();
+            $token = $session->getSessionVar('token');
+
+            Validation::logout();
+            if (isset($settingsJson) && isset($lastProvider)) {
+                $providerList = json_decode($settingsJson, true)['provider'];
+                $settings = $providerList[$lastProvider];
+
+                if (isset($settings) && key_exists('logoutUrl', $settings) && !empty($settings['logoutUrl']) && key_exists('clientId', $settings)) {
+                    $request->redirectUrl(
+                        $settings['logoutUrl'].
+                        '?client_id='.$settings['clientId'].
+                        // '&redirect_uri='.urlencode($router->url($request, $context->getPath(), "index").'/')
+                        '&post_redirect_uri='.$router->url($request, $context->getPath(), "index").
+                        '&id_token_hint='.$token
```

Comment:
I'm tempted to suggest adding `htmlspecialchars` wrappers around any string that gets included in a URL, when assembling URLs by concatenation.

Reply:
Changed it to 

```
$request->redirectUrl(
                                    $settings['authUrl'] .
                                    '?client_id=' . urlencode($settings['clientId']) .
                                    '&response_type=code' .
                                    '&scope=openid' .
                                    '&redirect_uri=' . urlencode($router->url($request, null, "openid", "doAuthentication", null, ['provider' => $name]))
                                );
```

here and elsewhere